Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp

```diff
@@ -384,19 +384,23 @@ namespace AzToolsFramework::ComponentModeFramework
     }
 
     // this is called twice when a component is added, once while the component is pending and
-    // once when it has been added fully to the entity. This is handled in UpdateComponentButton
+    // once when it has been added fully to the entity. 
     void ComponentModeSwitcher::OnEntityCompositionChanged(const AzToolsFramework::EntityIdList& entityIdList)
     {
         if (AZStd::ranges::find(entityIdList, m_componentModePair.GetEntityId()) != entityIdList.end())
         {
-            if (m_addOrRemove == AddOrRemoveComponent::Add)
+            if (m_addOrRemove == AddOrRemoveComponent::Remove)
             {
-                AddComponentButton(m_componentModePair);
+                ClearSwitcher();
             }
-            else
+            else if (m_addOrRemove == AddOrRemoveComponent::Add && m_callCount < 1)
             {
-                RemoveComponentButton(m_componentModePair);
+                m_callCount += 1;
+                return;
             }
+
+            UpdateSwitcher(entityIdList, {});
+            m_callCount = 0;
```

Comment:
Is there an alternate way to handle this logic that doesn't make as many assumptions about its calling pattern? This seems pretty fragile because any change to the order or frequency of calls happening somewhere else in the code can cause this to break.

Ideally, it seems like all composition changes could just perform logic similar to UpdateSwitcherOnEntitySelectionChange() where it clears everything out and rebuilds the button list based on the set of common components that now exist, and not need to track the number of calls or whether or not an addOrRemove happened.

Reply:
I was thinking to do that but I don't want the switcher to have to clear and repopulate the switcher everytime something happens, but do you think that has a relatively low overhead?

Reply:
In a general sense, I'm trying to push for correctness and maintainability over speed of execution here. The current design is maintaining and merging partial state knowledge ("are we adding or removing"? "how many times are we called"?) across separate API entry points, which seems like it's opening the door to errors and isn't _necessarily_ making things perceptibly faster, though I haven't measured the system to say concretely one way or the other.

Instead of trying to maintain partial state, I'm wondering if it's possible to just react to whatever state has currently changed (ideally in an efficient way), and not worry as much about reacting multiple times to multiple rapid state changes.

At a high level, it seems like at any point in time, the Switcher has a set of active entities and components that it is displaying, and it receives notifications that something has changed about one or more of those entities and components. So instead of trying to defer processing on the notifications, it seems like we just want an efficient way to process changes. I think these just boil down to something like this:
* Component added: Check to see if this component is already displayed. If not, check to see if it *should* be displayed by seeing if it exists on all selected entities and has a component mode. If so, then add it.
* Component removed: Check to see if this component is already displayed. If yes, then remove it.
* Entity changed: For each component on the entity, go through component added. For each component on the switcher, if it's not on the entity, remove it.

And then the Switcher would only send out change notifications if components have *actually* been changed in the Switcher list. With these changes, I think both m_callCount and m_addOrRemove could get removed entirely.

It's certainly possible that this could cause performance problems by performing some redundant logic, and it would be interesting to profile selection and modification of 1000 duplicate entities with several editable components on them, to see if there _are_ in fact any performance problems here. My _intuition_ is that if any performance problems exist, the solution to them would either be faster ways to merge the entity and component lists, or triggering less updates at a higher level, rather than trying to work around extra updates inside the Switcher.

But right now, I'm just asking this all more as a question. If you've tried alternatives, and measured this already, and this is the best way to optimize it, that's totally OK to say so. My request in that case would just be to add some super explicit comments in the code to this effect. :)

Reply:
I agree, the checking of the amount of calls is not necessarily sustainable especially if it changes to get called once for example in above example, it previously just accepted that it was being called twice and dealt with it so I can just revert that. With the switcher I guess there will only be a max of 10s of components for the foreseeable so performance wouldn't be negatively impacted. 

With regards to m_addOrRemove, when I receive a notification OnEntityComponentAdded originally I wanted to process the component then but the component doesn't "exist" on the entity until I receive a notification from OnEntityCompositionChanged which is why they are paired and share state knowledge. I don't like it but have tried searching for alternatives for getting the absolute state of if a component has been added or removed but couldn't find any.

I appreciate your feedback thank you :)